snapshots/snapshot_generator/src/main/scala/uk/ac/wellcome/platform/snapshot_generator/services/SnapshotService.scala

```diff
@@ -43,60 +51,100 @@ class SnapshotService(akkaS3Settings: S3Settings,
-        bucketName = publicBucketName,
-        objectKey = publicObjectKey
+      snapshotResult = SnapshotResult(
+        indexName = elasticConfig.worksIndex.name,
+        documentCount = documentCount,
+        displayModel = Work.getClass.getCanonicalName,
+        startedAt = startedAt,
+        finishedAt = Instant.now(),
+        s3Etag = uploadResult.etag,
+        s3Size = objectMetadata.contentLength,
+        s3Location = snapshotJob.s3Location
       )
 
+    } yield
       CompletedSnapshotJob(
         snapshotJob = snapshotJob,
-        targetLocation = targetLocation
+        snapshotResult = snapshotResult
       )
-    }
   }
 
-  private def runStream(publicBucketName: String,
-                        publicObjectKey: String,
-                        index: Index,
-                        toDisplayWork: Work.Visible[Identified] => DisplayWork)
-    : Future[MultipartUploadResult] = {
+  private def getObjectMetadata(
+    location: S3ObjectLocation): Future[ObjectMetadata] = {
+    val objectMetadataFlow = S3
+      .getObjectMetadata(bucket = location.bucket, key = location.key)
+      .withAttributes(S3Attributes.settings(akkaS3Settings))
+      .map {
+        case Some(objectMetadata) => objectMetadata
+        case None =>
+          val runtimeException = new RuntimeException(
+            s"No object found at $location"
+          )
+
+          error(
+            msg = "Failed getting ObjectMetadata!",
+            t = runtimeException
+          )
+
+          throw runtimeException
+      }
+
+    objectMetadataFlow.runWith(Sink.head)
+  }
+
+  private def uploadSnapshot(
+    location: S3ObjectLocation,
+    index: Index,
+    toDisplayWork: Work.Visible[Identified] => DisplayWork)
+    : (Future[Int], Future[MultipartUploadResult]) = {
 
-    // This source outputs DisplayWorks in the elasticsearch index.
     val displayWorks: Source[DisplayWork, Any] =
       ElasticsearchWorksSource(elasticClient = elasticClient, index = index)
         .via(IdentifiedWorkToVisibleDisplayWork(toDisplayWork))
 
-    // This source generates JSON strings of DisplayWork instances, which
-    // should be written to the destination snapshot.
-    val jsonStrings: Source[String, Any] = displayWorks
-      .via(DisplayWorkToJsonStringFlow.flow)
-
-    // This source generates gzip-compressed JSON strings, corresponding to
-    // the DisplayWork instances from the source snapshot.
-    val gzipContent: Source[ByteString, Any] = jsonStrings
-      .via(StringToGzipFlow())
-
     val s3Sink: Sink[ByteString, Future[MultipartUploadResult]] =
       S3.multipartUpload(
-          bucket = publicBucketName,
-          key = publicObjectKey
+          bucket = location.bucket,
+          key = location.key
         )
         .withAttributes(S3Attributes.settings(akkaS3Settings))
 
-    gzipContent.runWith(s3Sink)
+    val countingSink: Sink[DisplayWork, Future[Int]] =
+      Sink.fold[Int, DisplayWork](0)((acc, _) => acc + 1)
```

Comment:
nit: I would say `total` instead of `acc` here. I know it's the accumulated result, but let's be clear that it's a total count.